README.rst
.. image:: slytherlisp.png

SlytherLisp
===========

.. This README is in reStructuredText format. If you have Docutils installed,
   you can validate the format and make an HTML copy by typing:
     $ rst2html README.rst >README.html

.. Replace with your own names and Mines email addresses.

:Implemented By:
   * Partner One <partnerone@Mines.EDU>
   * Partner Two <partnertwo@Mines.EDU>
   * Partner Three <partnerthree@Mines.EDU>

.. Replace with the current deliverable you are working on. For example, if you
   are submitting the first deliverable, this should be 1 (as shown).

:Deliverable: 1

SlytherLisp is a lexically-scoped Scheme-like Lisp dialect for CSCI-400. Lisp
dialects use s-expressions as their primary syntax structure, just like we have
learned in class. For example, adding 3 and 4 is written as::

    (+ 3 4)

The ``+`` is the function we are calling, and ``3`` and ``4`` are the
arguments. This language only uses prefix notation, no infix or postfix
notation is used.

In SlytherLisp, there are two kinds of things we can call by placing in the
parentheses:

:Functions: Whose arguments are evaluated, and the function returns a value to
            substitute where it was written.
:Macros:    Whose arguments are not evaluated, and the macro returns an
            expression to fill in its place.

For example, the ``define`` macro defines functions::

   (define (f x y)
     (* (+ x y) y))

In this case, we defined a function named ``f`` which takes arguments ``x`` and
``y`` and returns ``(x + y) * y``. It is necessary for ``define`` to be a macro
since it needs the arguments unevaluated to create the definition.

Further examples of SlytherLisp code can be found in the ``examples``
directory. Also, as a beginner reference to Scheme and Lisp-like languages, the
`Structure and Interpretation of Computer Programs`__ (available *free* online)
book is recommended, even though SlytherLisp is not exactly the same as Scheme.

__ https://mitpress.mit.edu/sicp/full-text/book/book.html

Getting Started with the Starter Code
-------------------------------------

1. Open this ``README.rst`` file in a text editor and make changes to the name
   and email up top. Save the file before proceeding to step 2.

2. Install an editable copy of the application::

      $ python3.6 -m pip install -e . --user

   or::

      $ python3.7 -m pip install -e . --user

   - ``-e`` says to install an editable copy. If you omit this, you'll need to
     re-install every time you make changes.

   - ``--user`` says to install for the local user only. This way you do not
     need to use ``sudo``. The binaries should end up in something like
     ``$HOME/.local/bin``: make sure this is in your ``PATH``.

3. After installing the application, you should have access to the ``slyther``
   program anywhere on your system. Confirm you can run the ``slyther``
   command (even though it may result in a ``NotImplementedError``). Check
   ``slyther --help`` for potential command line arguments.

   - If you are still getting ``slyther: command not found``, please ask on
     Piazza or come to office hours for help.

4. Open up the files in the ``slyther`` directory and familiarize yourself with
   their structure.

   - The ``raise NotImplementedError(...)`` lines are for you to replace with
     working code. Typically, they will state which deliverable you need to
     complete them for.

   - The functions typically have a description of how they should work at the
     top in a docstring. This docstring usually has doctests in it too: that
     is, those lines that start with ``>>>`` in the docstring are actually unit
     tests as well! You are free to change these docstrings or remove as you
     please.

   - You are free to implement helper functions, etc., as you need: in fact,
     this project would be very hard to do without doing so.

   - If you wish to change how a particular interface works in the application,
     leave the original interface in, and have that call the modified
     interface. This is how (you, and I) can test your code even if you change
     the application structure.

5. Start coding!

Running Tests
-------------

To run the tests, type ``pytest`` from the base directory (where ``README.rst``
is located). You will need to specify at least one of four flags: ``--d1``,
``--d2``, ``--d3``, or ``--d4``, which runs the tests for Deliverable 1, 2, 3,
or 4. For example,

Running D1 tests::

   $ pytest --d1

Running all tests::

   $ pytest --d1 --d2 --d3 --d4

.. warning::

   Your grade is not solely based on unit tests. The instructor or grader
   *will* perform additional testing on your code.

Pretty Formatting
~~~~~~~~~~~~~~~~~

If you wish to show unit test results in your browser, a plugin for ``pytest``
is available.

https://pypi.org/project/pytest-html/

I highly recommend this plugin to help make looking at the test results easier.

Style Checking
--------------

The file ``.flake8`` defines a set of style checks (PEP 8, plus a few others).
To run the style checks, type ``flake8`` from the base directory. This is the
same way that the code style will be checked when graded.

With ``flake8``, no news is good news as well. If you want to make sure it's
working, add some bad style to your code for a second and see if it errors at
you.

Submitting your Deliverables
----------------------------

1. Make sure all code for your submission is completed.

2. Run the ``make_submission.sh`` script.

3. Upload the resulting ``submission.tar.bz2`` to Canvas.


history.txt

# 2019-05-04 13:06:52.591891
+cons

# 2019-05-04 13:07:14.523767
+(cons 1 2)

# 2019-05-05 11:42:33.037868
+(cons 12 32 )

# 2019-05-05 12:41:11.848863
+(match '(1 2 3 . 4)
+    [(list-rest a ... d) (list a d)])

# 2019-05-05 12:43:37.686336
+(print
+  "Your rating is:"
+  (cond
+    ((<= 3 18.5) 'UNDERWEIGHT)
+    ((<= 18 25) 'NORMAL)
+    ((<= 20 30) 'OVERWEIGHT)
+    (#t 'OBESE)))
+

# 2019-05-05 12:44:51.360341
+(list a ..3)

# 2019-05-05 12:45:06.775944
+(list a)

# 2019-05-06 14:27:13.871371
+'(1 _ 3)

# 2019-05-06 14:27:21.033524
+'(1 .. 3)

# 2019-05-06 14:27:27.485868
+'(1 ..3)

# 2019-05-06 14:27:55.441366
+'(1 .3)

# 2019-05-06 14:28:02.460987
+'(1 ... 3)

# 2019-05-06 14:33:00.091808
+(list 1 ..3)

# 2019-05-08 11:41:49.565927
+(list frequency ('(12 2 3) '(1 2))

# 2019-05-08 11:41:55.177961
+(list frequency ('(12 2 3) '(1 2)))

# 2019-05-08 11:43:18.633307
+(list_frequency ('(12 2 3) '(1 2)))

# 2019-05-08 11:43:39.728443
+(list_frequency '('(12 2 3) '(1 2)))

# 2019-05-08 11:43:54.064451
+(list_frequency '(12 2 3) '(1 2))

# 2019-05-08 11:44:08.080602
+'(1 ..3)

# 2019-05-08 11:44:49.187946
+(list_frequency '(12 2 3) '(1 2))

# 2019-05-08 12:49:35.555766
+(list_frequency '(12 2 3 3 2 12) '(12 2 3))

# 2019-05-08 12:53:23.498955
+'(1 ..3)

# 2019-05-08 12:56:24.647378
+(list_frequency '(12 2 3 3 2 12) '(12 2 3))

# 2019-05-08 13:04:22.858558
+(list_frequency '(12 2 3 3 2 12) '(12 2 3 4))

# 2019-05-08 13:04:28.991032
+(list_frequency '(12 2 3 3 2 12) '(12 2 3))

# 2019-05-08 13:06:04.034885
+(list_frequency '(12 2 3 3 2 12) '(12 2 3 4))

# 2019-05-08 13:16:05.637437
+(list_frequency '(12 2 3 3 2 12) '(12 2 3))

# 2019-05-08 13:16:54.174314
+(list_frequency '(12 2 3 3 2 12) '(12 2 3 4))

# 2019-05-08 13:18:02.987982
+(list_frequency '(12 2 3 3 2 12) '(12 2 3))

# 2019-05-08 13:18:43.743271
+(list_frequency '(12 2 3 3 2 12) '(12 2 3 4))

# 2019-05-08 13:18:56.543083
+(list_frequency '(12 2 3 3 2 3 2 2 22 12 12) '(12 2 3 4))

# 2019-05-08 13:19:02.277549
+(list_frequency '(12 2 3 3 2 3 2 2 22 12 4 12) '(12 2 3 4))

# 2019-05-08 14:22:43.821984
+(list_frequency '(12 2 3 3 2 3 2 2 22 12 4 12) '())

# 2019-05-08 14:22:54.545667
+(list_frequency '() '(12 2 3 4))

# 2019-05-08 14:23:45.336675
+(list_frequency '(12 2 3 3 2 3 2 2 22 12 4 12) '())

# 2019-05-08 14:23:52.402182
+(list_frequency '(12 2 3 3 2 3 2 2 22 12 4 12) '(12 2 3 4))

# 2019-05-08 14:23:59.856433
+(list_frequency '() '(12 2 3 4))

# 2019-05-08 14:26:52.218161
+(list_frequency '() '(12 2 3 4 5))

# 2019-05-10 20:36:39.565092
+(plot?)

# 2019-05-10 20:37:16.743226
+(plot-scatter '(1 2 3 4 5) '(6 7 8 9 10) "--p")

# 2019-05-10 20:38:26.029572
+(plot-pie '("Java" "C++" "Python") '(98 88 30) "--p")

# 2019-05-10 20:41:10.257855
+(plot-pie '("C++" "Java" "Python") '(98 88 50) "--p")

# 2019-05-10 20:41:58.840315
+(plot-pie '("Java" "C++" "Python") '(98 88 30) "--p")

# 2019-05-10 20:53:49.120952
+(plot-scatter '(1 2 3 4) '(5 6 7 8) "--p")

# 2019-05-10 20:54:09.121160
+(plot-pie '("Java" "C++" "Python") '(98 88 30) "--p")

# 2019-05-10 21:01:59.531820
+(plot-scatter '(1 2 3 4) '(5 6 7 8) "--p")


tests/conftest.py
import pytest


deliverables = ['d1', 'd2', 'd3', 'd4']


def pytest_addoption(parser):
    for d in deliverables:
        parser.addoption(
            '--' + d,
            action='store_true',
            help='Run tests for {}'.format(d.upper()))


@pytest.hookimpl(trylast=True)
def pytest_collection_modifyitems(config, items):
    from os import sep
    orig = items[:]
    items.clear()
    for item in orig:
        d = item.location[0].split(sep)[-2]
        if ((d in deliverables and config.getoption(d))
                or d not in deliverables):
            items.append(item)


slyther/repl.py
from prompt_toolkit import prompt
from prompt_toolkit.history import FileHistory
from prompt_toolkit.auto_suggest import AutoSuggestFromHistory
from prompt_toolkit.completion import WordCompleter

def repl(interpreter, debug=False):
    """
    Take an interpreter object (see ``slyther/interpreter.py``) and give a REPL
    on it. Should not return anything: just a user interface at the terminal.
    For example::

        $ slyther
        > (print "Hello, World!")
        Hello, World!
        NIL
        > (+ 10 10 10)
        30

    When the user presses ^D at an empty prompt, the REPL should exit.

    When the user presses ^C at any prompt (whether there is text or
    not), the input should be cancelled, and the user prompted again::

        $ slyther
        > (blah bla^C
        >                   <-- ^C resulted in new prompt line

    Should be pretty easy. No unit tests for this function, but I will
    test the interface works when I grade it.

    Optionally, you may want to prevent the REPL from raising an exception
    when an exception in user code occurs, and allow the user to keep
    typing further expressions. This is not required, just a suggestion.
    If you do this, you should probably disable this behavior when ``debug``
    is set to ``True``, as it allows for easy post-mortem debugging with pdb
    or pudb.
    """
    # Word completer from slyther.builtins.py
    slyther_completer = WordCompleter([
        'abs', 'and', 'car', 'cdr', 'ceil', 'cond', 'cons', 'define', 'eval', 'expt', 'floor', 'floordiv', 'if', 'input', 'lambda', 'let', 'list', 'make-float', 'make-integer', 'make-string', 'make-symbol', 'not', 'or', 'remainder', 'parse', 'print', 'set!', 'sqrt'], ignore_case=True)

    while True:
        try:
            expr = prompt('>',
                          history=FileHistory('history.txt'),
                          auto_suggest=AutoSuggestFromHistory(),
                          completer=slyther_completer)
            print(interpreter.exec(expr))
        except KeyboardInterrupt:
            print("")
            continue
        except EOFError:
            exit(0)


examples/list_freq.scm
#!/usr/bin/env slyther
(define list-1 '(12 3 4 5 6 8 19 202 30 40 3 3))
(define list-2 '("hello" "hello" "world" "this" "is" "d6" "lisp_eval"
                 "lisp"))
(define list-3 '())
(define list-4 '(#t #t #t #f #f #t #f))

(define target-1 '(12 3 4 5))
(define target-2 '("hello" "world" "lisp"))
(define target-3 '(12 3 4 5))
(define target-4 '(#t #f))
(define target-5 '())

(print "for '(12 3 4 5 6 8 19 202 30 40 3 3) as input and target '(12 3 4 5)")
(print "output is : ")
(print (list-frequency list-1 target-1))
(print "should be : (list (list 12 1) (list 3 3) (list 4 1) (list 5 1))\n")

(print "for '(\"hello\" \"hello\" \"world\" \"this\" \"is\" \"d6\" \"lisp_eval\"\"lisp\") as input and target '(\"hello\" \"world\" \"lisp\")")
(print "output is : ")
(print (list-frequency list-2 target-2))
(print "should be : (list (list \"hello\" 2) (list \"world\" 1) (list \"lisp\" 1))\n")

(print "for '() as input and target  '(12 3 4 5)")
(print "output is : ")
(print (list-frequency list-3 target-3))
(print "should be : (list (list 12 0) (list 3 0) (list 4 0) (list 5 0))\n")

(print "for '(#t #t #t #f #f #t #f) as input and target  '(#t #f)")
(print "output is : ")
(print (list-frequency list-3 target-4))
(print "should be : (list (list #t 4) (list #f 3))\n")

(print "for '(12 3 4 5 6 8 19 202 30 40 3 3) as input and target  '()")
(print "output is : ")
(print (list-frequency list-3 target-5))
(print "should be : NIL\n")

slyther/builtins.py
import operator
from functools import reduce
from slyther.types import (BuiltinFunction, BuiltinMacro, Symbol,
                           UserFunction, SExpression, cons, String,
                           Variable, ConsList, NIL, LexicalVarStorage,
                           ConsCell)
from slyther.evaluator import lisp_eval
from slyther.parser import lex, parse, lisp
from math import floor, ceil, sqrt


@BuiltinFunction('+')
def add(*args):
    """
    Sum each of the arguments

    >>> add(1, 2, 3, 4)
    10
    >>> add()
    0

    .. hint::

        Use ``sum`` or ``reduce``.
    """
    return sum(args)


@BuiltinFunction('-')
def sub(*args):
    """
    ``(- x y z)`` computes ``x - y - z``, but with only one argument,
    ``(- x)`` computes ``-x``.

    >>> sub(10, 1, 2)
    7
    >>> sub(10, 1)
    9
    >>> sub(10)
    -10
    >>> sub()
    0

    .. hint::

        Use ``reduce``.
    """
    if len(args) == 1:
        return -1 * args[0]
    if len(args) > 1:
        return reduce(lambda x, y: x - y, args)
    else:
        return 0


@BuiltinFunction('*')
def mul(*args):
    """
    Compute the product.

    >>> mul(2, 3)
    6
    >>> mul(2, 3, 3)
    18
    >>> mul(2)
    2
    >>> mul()
    1
    """
    product = 1
    for i in args:
        product *= i
    return product


@BuiltinFunction('/')
def div(*args):
    """
    ``(/ a b c)`` computes ``a / b / c``, but ``(/ a)`` computes
    ``1 / a``.

    >>> div(1, 2)
    0.5
    >>> div(1, 2, 2)
    0.25
    >>> div(2)
    0.5
    """
    if len(args) == 1:
        return operator.truediv(1, args[0])
    if len(args) > 1:
        return reduce(lambda x, y: operator.truediv(x, y), args)
    else:
        return 0


@BuiltinFunction
def floordiv(*args):
    """
    Equivalent to ``div``, but uses ``operator.floordiv``.

    >>> floordiv(3, 2)
    1
    >>> floordiv(3, 2, 2)
    0
    >>> floordiv(1)
    1
    >>> floordiv(2)
    0
    """
    if len(args) == 1:
        return operator.floordiv(1, args[0])
    if len(args) > 1:
        return reduce(lambda x, y: operator.floordiv(x, y), args)
    else:
        return 0


# IO
print_ = BuiltinFunction(print)
input_ = BuiltinFunction(input)

# Type Constructors
make_int = BuiltinFunction(int, 'make-integer')
make_float = BuiltinFunction(float, 'make-float')
make_symbol = BuiltinFunction(Symbol, 'make-symbol')
make_string = BuiltinFunction(String, 'make-string')


@BuiltinFunction('list')
def list_(*args) -> ConsList:
    """
    Create a ``ConsList`` from ``args``.

    >>> list_(1, 2, 3)
    (list 1 2 3)
    >>> list_()
    NIL
    """
    return ConsList.from_iterable(args)


# Comparators
lt = BuiltinFunction(operator.lt, '<')
gt = BuiltinFunction(operator.gt, '>')
eq = BuiltinFunction(operator.eq, '=')
le = BuiltinFunction(operator.le, '<=')
ge = BuiltinFunction(operator.ge, '>=')
not_ = BuiltinFunction(operator.not_, 'not')

# arithmetic
remainder = BuiltinFunction(operator.mod, 'remainder')
floor_ = BuiltinFunction(floor)
ceil_ = BuiltinFunction(ceil)
sqrt_ = BuiltinFunction(sqrt)
abs_ = BuiltinFunction(abs)
expt = BuiltinFunction(operator.pow, 'expt')

# string manipulation
format_ = BuiltinFunction(str.format)
split = BuiltinFunction(str.split)

# cons cell functions
cons = BuiltinFunction(cons)


@BuiltinFunction
def car(cell: ConsCell):
    """
    Get the ``car`` of a cons cell.
    """
    return cell.car


@BuiltinFunction
def cdr(cell: ConsCell):
    """
    Get the ``cdr`` of a cons cell.
    """
    return cell.cdr


@BuiltinFunction('nil?')
def is_nil(cell: ConsCell) -> bool:
    """
    Return ``True`` if the cell is ``NIL``, ``False`` otherwise.
    """
    return cell.car == NIL


@BuiltinMacro
def define(se: SExpression, stg: LexicalVarStorage):
    """
    Define a variable or a function to a value, calling ``put`` on
    the storage::

        ; variable
        (define var-name evaluate-me)

        ; lambda function
        (define func-name (lambda (args...) (body1) ... (bodyN)))

        ; SE-named function
        (define (func-name args...) (body1) ... (bodyN))

    .. note::

        If defining a function (either by ``lambda`` or by SE-named
        syntax), the definition **must** be made visible from within
        that function's ``environ``, or recursion will not work!

    >>> from slyther.types import *
    >>> from slyther.parser import lisp
    >>> se = lisp('((twirl alpha beta) (print alpha) (print beta))')
    >>> name = Symbol('twirl')
    >>> args = lisp('(alpha beta)')
    >>> body = lisp('((print alpha) (print beta))')
    >>> stg = LexicalVarStorage({})
    >>> stg.put('NIL', NIL)
    >>> stg.put('define', define)
    >>> stg.put('lambda', lambda_func)
    >>> from slyther.evaluator import lisp_eval
    >>> lisp_eval(define(cons(cons(name, args), body), stg), stg)
    NIL
    >>> lisp_eval(define(lisp('(x 10)'), stg), stg)
    NIL
    >>> stg[name].value
    (lambda (alpha beta) (print alpha) (print beta))
    >>> stg[name].value.environ['twirl'].value
    (lambda (alpha beta) (print alpha) (print beta))
    >>> stg['x'].value
    10
    >>> stg[name].value.environ['x'].value
    Traceback (most recent call last):
        ...
    KeyError: 'x'
    """
    key = se.car
    value = se.cdr

    if isinstance(key, SExpression):
        function = UserFunction(params=key.cdr, body=value, environ=stg.fork())
        key = key.car
        function.environ[key] = Variable(function)
        stg.put(key, function)
    elif isinstance(key, Symbol):
        stg.put(key, lisp_eval(value.car, stg))
    else:
        stg.put(key, value)


@BuiltinMacro('lambda')
def lambda_func(se: SExpression, stg: LexicalVarStorage) -> UserFunction:
    """
    Define an anonymous function and return the ``UserFunction``
    object.

    >>> from slyther.types import *
    >>> stg = LexicalVarStorage({})
    >>> stg.put('x', 20)
    >>> f = lambda_func(
    ...     SExpression.from_iterable([
    ...         SExpression.from_iterable(map(Symbol, ['a', 'b', '.', 'c'])),
    ...         SExpression.from_iterable(map(Symbol, ['print', 'a'])),
    ...         SExpression.from_iterable(map(Symbol, ['print', 'b'])),
    ...         SExpression.from_iterable(map(Symbol, ['print', 'c']))]),
    ...     stg)
    >>> f
    (lambda (a b . c) (print a) (print b) (print c))
    >>> f.environ['x'].value
    20
    """
    return UserFunction(se.car, se.cdr, stg.fork())


@BuiltinMacro('let')
def let(se: SExpression, stg: LexicalVarStorage) -> SExpression:
    """
    The ``let`` macro binds variables to a local scope: the expressions
    inside the macro. Like a function, a ``let`` returns the last
    expression in its body. Once the ``let`` returns, the variables are
    unbound, and cannot be accessed anymore. For example::

        (let ((a (+ f g))
              (b 11)
              (c 12))
          (+ a (- b c)))

    In the example above, ``a`` was bound to whatever ``(+ f g)``
    evaluates to, b to ``11``, and ``c`` to ``12``.

    Notice how the above was equivalent to another expression::

        ((lambda (a b c)
           (+ a (- b c))) (+ f g) 11 12)

    You should do a syntax translation from the input to something like that.

    >>> from slyther.types import *
    >>> from slyther.parser import lisp
    >>> from slyther.evaluator import lisp_eval
    >>> stg = LexicalVarStorage(
    ...         {'let': Variable(let),
    ...          'lambda': Variable(lambda_func),
    ...          'print': Variable(print_),
    ...          'x': Variable(10)})
    >>> lisp_eval(lisp('(let ((x 20) (y 30)) (print x) (print y))'), stg)
    20
    30
    NIL
    >>> lisp_eval(Symbol('x'), stg)
    10
    """
    vals = []
    ptrs = []
    pred = se.car
    con = se.cdr
    for item in pred:
        vals.append(item.car)
        ptrs.append(item.cdr.car)
    param = SExpression.from_iterable(vals)
    ptrs = SExpression.from_iterable(ptrs)
    function = UserFunction(params=param, body=con, environ=stg.fork())
    res = SExpression(function, ptrs)

    return res


@BuiltinMacro('if')
def if_expr(se: SExpression, stg: LexicalVarStorage):
    """
    An ``if`` expression looks like this::

        (if <predicate> <consequent> <alternative>)

    If the predicate evaluates to something truthy, return the
    consequent, otherwise return the alternative. An example::

        (if (< x 10)
          (print "x is less than 10")
          (print "x is greater than or equal to 10"))

    >>> from slyther.types import *
    >>> from slyther.parser import lisp
    >>> se = lisp('((< x 10)'
    ...           ' (print "x is less than 10")'
    ...           ' (print "x is greater than or equal to 10"))')
    >>> stg = LexicalVarStorage({})
    >>> stg.put('<', lt)
    >>> stg.put('x', 9)
    >>> if_expr(se, stg)
    (print "x is less than 10")
    >>> stg['x'].set(10)
    >>> if_expr(se, stg)
    (print "x is greater than or equal to 10")
    """
    if lisp_eval(se.car, stg):
        return se.cdr.car
    else:
        return se.cdr.cdr.car


@BuiltinMacro('cond')
def cond(se: SExpression, stg: LexicalVarStorage):
    """
    ``cond`` is similar to ``if``, but it lists a series of predicates
    and consequents, similar to how guards work in Haskell. For
    example::

        (cond
          ((< x 5) (print "x < 5"))
          ((< x 10) (print "5 <= x < 10"))
          ((< x 15) (print "10 <= x < 15"))
          (#t (print "x >= 15")))

    >>> from slyther.types import *
    >>> from slyther.parser import lex, parse
    >>> def test_cond(x):
    ...     expr = next(parse(lex('''
    ...         (cond
    ...           ((< x 5) (print "x < 5"))
    ...           ((< x 10) (print "5 <= x < 10"))
    ...           ((< x 15) (print "10 <= x < 15"))
    ...           (#t (print "x >= 15")))''')))
    ...     stg = LexicalVarStorage({})
    ...     stg.put('<', lt)
    ...     stg.put('#t', Boolean(True))
    ...     stg.put('x', x)
    ...     return cond(expr.cdr, stg)
    >>> test_cond(4)
    (print "x < 5")
    >>> test_cond(5)
    (print "5 <= x < 10")
    >>> test_cond(10)
    (print "10 <= x < 15")
    >>> test_cond(15)
    (print "x >= 15")
    """
    while se is not NIL:
        if(lisp_eval(se.car.car, stg)):
            return se.car.cdr.car
        se = se.cdr


@BuiltinMacro('and')
def and_(se: SExpression, stg: LexicalVarStorage):
    """
    Compute an ``and`` expression, like this::

        (and (< x 10) (> y 15) (foo? z))

    Evaluate left to right, and return the first result which produces
    a falsy value. Note that the result need not be a boolean, but you
    should test its falsiness(parse(lex('''
    ...         (cond
    ...           ((< x 5) (print "x < 5"))
    ...           ((< x 10) (print "5 <= x < 10"))
    ...           ((< x 15) (print "10 <= x < 15"))
    ...           (#t (print "x >= 15")))'''))), and return the result (even if it's not
    a boolean).

    Note that you could return the last expression unevaluated if all
    the previous are truthy, as your ``lisp_eval`` should eval it for
    you. This could be useful for tail call optimization.

    >>> from slyther.types import *
    >>> from slyther.parser import lisp
    >>> from slyther.evaluator import lisp_eval
    >>> stg = LexicalVarStorage(
    ...         {'and': Variable(and_),
    ...          '#f': Variable(Boolean(False)),
    ...          '#t': Variable(Boolean(True)),
    ...          'print': Variable(print_),
    ...          '+': Variable(add),
    ...          'x': Variable(0),
    ...          'y': Variable(1),
    ...          'z': Variable(2)})
    >>> lisp_eval(lisp('(and #t (print (+ x y z)) #f)'), stg)
    3
    NIL
    >>> lisp_eval(lisp('(and #f (print (+ x y z)) #f)'), stg)
    #f
    >>> lisp_eval(lisp('(and #t x (print (+ x y z)) y foo)'), stg)
    0
    >>> lisp_eval(lisp('(and (+ x y) y z)'), stg)
    2
    >>> lisp_eval(lisp('(and)'), stg)
    NIL
    """
    res = NIL
    length = len(se)
    for index, x in enumerate(se):
        if index == length - 1:
            return x
        y = lisp_eval(x, stg)
        res = y
        if not y:
            return y
    return res


@BuiltinMacro('or')
def or_(se: SExpression, stg: LexicalVarStorage):
    """
    Similar to ``and`` above, but compute an ``or`` instead. Return
    the first truthy value rather than falsy.

    >>> from slyther.types import *
    >>> from slyther.parser import lisp
    >>> from slyther.evaluator import lisp_eval
    >>> stg = LexicalVarStorage(
    ...         {'or': Variable(or_),
    ...          '#f': Variable(Boolean(False)),
    ...          '#t': Variable(Boolean(True)),
    ...          'print': Variable(print_),
    ...          '+': Variable(add),
    ...          'x': Variable(0),
    ...          'y': Variable(1),
    ...          'z': Variable(2)})
    >>> lisp_eval(lisp('(or #t (print (+ x y z)) #f)'), stg)
    #t
    >>> lisp_eval(lisp('(or #f (print (+ x y z)) #f)'), stg)
    3
    #f
    >>> lisp_eval(lisp('(or #f x (print (+ x y z)) y foo)'), stg)
    3
    1
    >>> lisp_eval(lisp('(or (+ y -1) x z)'), stg)
    2
    >>> lisp_eval(lisp('(or)'), stg)
    NIL
    """
    res = NIL
    length = len(se)
    for index, x in enumerate(se):
        if index == length - 1:
            return x
        y = lisp_eval(x, stg)
        res = y
        if y:
            return y
    return res

@BuiltinMacro('set!')
def setbang(se: SExpression, stg: LexicalVarStorage):
    """
    ``set!`` is the assigment macro. Unlike ``define``, ``set!`` **does not
    create new variables.** Instead, it changes the value of existing
    variables **only**. If given a variable which has not been defined,
    it should raise a ``KeyError``.

    For example::

        (set! <name> <value>)

    Should just eval ``value`` and call ``stg[name].set`` on it. Return
    ``NIL``.

    >>> from slyther.types import *
    >>> from slyther.parser import lisp
    >>> stg = LexicalVarStorage({'foo': Variable(12), 'bar': Variable(15)})
    >>> old_foo = stg['foo']
    >>> setbang(lisp('(foo 15)'), stg)
    NIL
    >>> stg['foo'].value
    15
    >>> stg['foo'] is old_foo
    True
    >>> setbang(lisp('(baz 15)'), stg)
    Traceback (most recent call last):
        ...
    KeyError: 'Undefined variable baz'
    """
    try:
        stg[se.car].set(lisp_eval(se.cdr.car, stg))
        return NIL
    except KeyError as ex:
        raise KeyError("Undefined variable {}".format(str(se.car))) from ex


@BuiltinMacro('eval')
def eval_(se: SExpression, stg: LexicalVarStorage):
    """
    ::

        (eval <expr>)

    Evaluate ``expr``, if it produces a ``ConsList``, then upgrade that
    ``ConsList`` to an ``SExpression`` (recursively) and return it. Your
    ``lisp_eval`` will take care of actually evaluating the expression,
    as this is a macro.

    >>> from slyther.types import *
    >>> from slyther.parser import lisp
    >>> from slyther.evaluator import lisp_eval
    >>> stg = LexicalVarStorage(
    ...         {'eval': Variable(eval_),
    ...          '#t': Variable(Boolean(True)),
    ...          'print': Variable(print_),
    ...          'x': Variable(0),
    ...          'y': Variable(1),
    ...          'z': Variable(lisp("(print x)"))})
    >>> lisp_eval(lisp("(eval '(print x))"), stg)
    0
    NIL
    >>> lisp_eval(lisp("(eval ''(print x))"), stg)
    (list print x)
    >>> lisp_eval(lisp('(eval #t)'), stg)
    #t
    >>> lisp_eval(lisp("(eval '''#t)"), stg)
    '#t
    >>> lisp_eval(lisp("(eval 'y)"), stg)
    1
    >>> lisp_eval(lisp("(eval ''y)"), stg)
    y
    >>> lisp_eval(lisp("(eval z)"), stg)
    0
    NIL
    """
    expression = lisp_eval(se.car, stg)
    if isinstance(expression, ConsList):
        expression = SExpression.from_iterable(
            eval_(SExpression(x), stg) for x in expression)

    return expression


@BuiltinFunction('parse')
def parse_string(code: String):
    """
    ``code`` is a ``String`` containing a single thing once parsed, such as an
    s-expression, or a symbol. Simply just lex it, parse it and return it!
    Woo-hoo!

    >>> from slyther.types import *
    >>> parse_string(String("(print x)"))
    (list print x)

    Note that the ``BuiltinFunction`` decorator takes care of downgrading an
    ``SExpression`` to a ``ConsList`` for you.
    """
    return next(parse(lex(code)))

# Given an input list, we will count the frequency from the members in the
# target list.
@BuiltinFunction('list-frequency')
def list_frequency(input_list: SExpression, target_list: SExpression):
    """
    $slyther
    >(list_frequency '(12 2 3 3 2 3 2 2 22 12 4 12) '(12 2 3 4))
    (list (list 12 3) (list 2 4) (list 3 3) (list 4 1))

    """
    
    # create an map that stores all the elements from input_list
    table = {}
    while input_list is not NIL:
        key = input_list.car
        if key in table:
            table[key] += 1
        else:
            table[key] = 1
        
        input_list = input_list.cdr

    root = ConsList(0, NIL)
    res = root
    check = 0
    size = len(target_list)

    if size == 0:
        return NIL

    # delete the keys that are not in the input list 
    while target_list is not NIL:
        check += 1
        key = target_list.car
        if key in table:
            value = table[key]
            root.car = ConsList(key, NIL)
            root.car.cdr = ConsList(value, NIL)
        else:
            root.car = ConsList(key, NIL)
            root.car.cdr = ConsList(0, NIL)
        if check != size:
            root.cdr = ConsList(0, NIL)
            root = root.cdr
        
        target_list = target_list.cdr
    
    # generate the result:

    return res


@BuiltinFunction('curry')
def curry(se: SExpression, stg: LexicalVarStorage=LexicalVarStorage({})):
    """
    curry returns a user function with given arguments,
    It will return the user function if it is a user function
    
    """
    if isinstance(se, UserFunction):
        return se

@BuiltinFunction('plot-scatter')
def plot_scat(x: ConsList, y:ConsList, option: String):
    import matplotlib.pyplot as plt
    """
    Plots to scatter_plot.pdf given arguments x y respectively
    """
    if option == String("--h"):
        print("Enter your data as follows given x and y \n (plot-scatter '(x1 .. xn) '(y1 ... yn) \"--p)\"\n to plot the graph to scatter_plot.pdf \n (plot-scatter '(x1 .. xn) '(y1 .. yn) \"--h\"\n for help")
    elif option == String("--p"):
        data_x = []
        data_y = []
        while x is not NIL:
            data_x.append(x.car)
            x = x.cdr
        while y is not NIL:
            data_y.append(y.car)
            y = y.cdr
        print("Your x axis data", data_x)
        print("Your y axis data", data_y)
        if len(data_x) != len(data_y):
            print("Your data lenghts do not match!")
        else:
            title = input("Enter the title for this plot: ")
            xlabel = input("Enter the xlabel for this plot: ")
            ylabel = input("Enter the ylabel for this plot: ")
            plt.scatter(data_x, data_y, c='r')
            plt.title(title)
            plt.xlabel(xlabel)
            plt.ylabel(ylabel)
            plt.savefig("scatter_plot.pdf")
            print("Your plot is saved as scatter_plot.pdf. Thank You for using Slyther plot!")
            
@BuiltinFunction('plot-pie')
def plot_pie(x: ConsList, y: ConsList, option: String):
    import matplotlib.pyplot as plt
    """
    Plots to pie_plot.pdf given arguments x y respectively
    """
    if option == String("--h"):
        print("Enter your data as follows given labels(x) and y \n (plot-scatter '(x1 .. xn) '(y1 ... yn) \"--p)\"\n to plot the graph to pie_plot.pdf \n (plot-scatter '(x1 .. xn) '(y1 .. yn) \"--h\"\n for help")
    elif option == String("--p"):
        data_x = []
        data_y = []
        while x is not NIL:
            data_x.append(x.car)
            x = x.cdr
        while y is not NIL:
            data_y.append(y.car)
            y = y.cdr
        print("Your x axis data", data_x)
        print("Your y axis data", data_y)
        if len(data_x) != len(data_y):
            print("Your data lenghts do not match!")
        else:
            title = input("Enter the title for this plot: ")
            patches, texts = plt.pie(data_y, startangle=90)
            plt.legend(patches, data_x, loc="best")
            plt.axis('equal')
            plt.tight_layout()
            plt.title(title)
            plt.savefig("pie_plot.pdf")
            print("Your plot is saved as pie_plot.pdf. Thank You for using Slyther plot!")

@BuiltinFunction('plot?')
def plot_help():
    print("To plot scatter plots, use plot-scatter function")
    print("To learn how to use that function you enter (plot-scatter '() '() \"--h\"")
    print("")
    print("To plot pie plots, use plot-pie function")
    print("To learn how to use the function you can enter (plot-scatter '() '() \"--h\"")
